python3.8/cookiecutter-aws-sam-step-functions-sample-app/{{cookiecutter.project_name}}/tests/integration/test_state_machine.py

```diff
@@ -0,0 +1,162 @@
+        )
+
+    def _start_execute(self):
+        """
+        Start the state machine execution request and record the execution ARN
+        """
+        response = self.client.start_execution(
+            stateMachineArn=self.state_machine_arn, name=f"integ-test-{uuid4()}", input="{}"
+        )
+        self.execution_arn = response["executionArn"]
+
+    def _wait_execution(self):
+        while True:
+            response = self.client.describe_execution(executionArn=self.execution_arn)
+            status = response["status"]
+            if status == "SUCCEEDED":
+                logging.info(f"Execution {self.execution_arn} completely successfully.")
+                break
+            elif status == "RUNNING":
+                logging.info(f"Execution {self.execution_arn} is still running, waiting")
+                sleep(3)
+            else:
+                self.fail(f"Execution {self.execution_arn} failed with status {status}")
+
+    def _retrieve_transaction_table_input(self):
+        """
+        Make sure "Record Transaction" step was reached, and record the input of it.
+        """
+        response = self.client.get_execution_history(executionArn=self.execution_arn)
+        events = response["events"]
+        record_transaction_entered_events = [
+            event
+            for event in events
+            if event["type"] == "TaskStateEntered" and event["stateEnteredEventDetails"]["name"] == "Record Transaction"
+        ]
+        self.assertTrue(
+            record_transaction_entered_events,
+            "Cannot find Record Transaction TaskStateEntered event",
+        )
+        self.transaction_table_input = json.loads(
+            record_transaction_entered_events[0]["stateEnteredEventDetails"]["input"]
+        )
+
+    def _verify_transaction_record_written(self):
+        """
+        Use the input recorded in _retrieve_transaction_table_input() to
+        verify whether the record has been written to dynamodb
+        """
+        client = boto3.client("dynamodb")
+        response = client.get_item(
+            Key={
+                "Id": {
+                    "S": self.transaction_table_input["id"],
+                },
+            },
+            TableName=self.transaction_table_name,
+        )
+        self.assertTrue(
+            "Item" in response,
+            f'Cannot find transaction record with id {self.transaction_table_input["id"]}',
+        )
+        item = response["Item"]
+        self.assertDictEqual(item["Quantity"], {"N": self.transaction_table_input["qty"]})
+        self.assertDictEqual(item["Price"], {"N": self.transaction_table_input["price"]})
+        self.assertDictEqual(item["Type"], {"S": self.transaction_table_input["type"]})
+
+    def test_state_machine(self):
+        self._start_execute()
+        self._wait_execution()
+        self._retrieve_transaction_table_input()
+        self._verify_transaction_record_written()
```

Comment:
I wonder is there a more idiomatic way of writing this? Maybe using return values. Currently it requires to keep track mentally of the attributes and the effects of each method.

Reply:
addressed